### PR TITLE
fix: migrate base driver to ts

### PR DESCRIPTION
**Pull request by frolovdev**
**Check List**
- [ ] Tests has been run in packages where changes made if available
- [ ] Linter has been run for changed code
- [ ] Tests for the changes have been added if not covered yet
- [ ] Docs have been added / updated if required

**Issue Reference this PR resolves**

[For example #12]

**Description of Changes Made (if issue reference is not provided)**

[Description goes here]

### CODE REVIEW

File: packages/cubejs-query-orchestrator/src/driver/driver.interface.ts [hunk 1]
```diff
@@ -109,7 +125,10 @@ export type QueryOptions = {
   inlineTables?: InlineTables,
   [key: string]: any
 };
-export type DownloadQueryResultsResult = DownloadQueryResultsBase & (DownloadTableMemoryData | DownloadTableCSVData | StreamTableData | StreamingSourceTableData);
+export type DownloadQueryResultsResult = DownloadQueryResultsBase & (DownloadTableMemoryData | DownloadTableCSVData | StreamTableData | StreamingSourceTableData | StreamTableDataWithTypes);
```
Comment (frolovdev) [inline packages/cubejs-query-orchestrator/src/driver/driver.interface.ts:128]: 📓 StreamTableDataWithTypes required in postgress streaming

File: packages/cubejs-query-orchestrator/src/driver/driver.interface.ts [hunk 2]
```diff
@@ -75,6 +87,10 @@ export type StreamTableDataWithTypes = StreamTableData & {
   types: TableStructure;
 };
 
+export function isDownloadTableMemoryData(tableData: any): tableData is DownloadTableMemoryData {
```
Comment (frolovdev) [inline packages/cubejs-query-orchestrator/src/driver/driver.interface.ts:90]: 📓 this function allows narrow the union

File: packages/cubejs-query-orchestrator/src/driver/BaseDriver.ts [hunk 3]
```diff
@@ -261,79 +237,72 @@ export class BaseDriver {
...
-          `INSERT INTO ${table}
-        (${columns.map(c => this.quoteIdentifier(c.name)).join(', ')})
-        VALUES (${columns.map((c, paramIndex) => this.param(paramIndex)).join(', ')})`,
-          columns.map(c => this.toColumnValue(tableData.rows[i][c.name], c.type))
-        );
-      }
-      for (let i = 0; i < indexesSql.length; i++) {
-        const [query, params] = indexesSql[i].sql;
-        await this.query(query, params);
+      if (isDownloadTableMemoryData(tableData)) {
```
Comment (frolovdev) [inline packages/cubejs-query-orchestrator/src/driver/BaseDriver.ts:278]: 📓 just added if to narrow the type

File: packages/cubejs-query-orchestrator/src/orchestrator/PreAggregations.ts [hunk 4]
```diff
@@ -906,7 +906,7 @@ export class PreAggregationLoader {
     this.logExecutingSql(queryOptions);
     this.logger('Downloading external pre-aggregation via query', queryOptions);
     const externalDriver = await this.externalDriverFactory();
-    const capabilities = externalDriver.capabilities && externalDriver.capabilities();
```
Comment (frolovdev) [inline packages/cubejs-query-orchestrator/src/orchestrator/PreAggregations.ts:909]: 📓 always exist

File: packages/cubejs-query-orchestrator/src/driver/BaseDriver.ts [hunk 5]
```diff
@@ -261,79 +237,72 @@ export class BaseDriver {
...
+        }
       }
     } catch (e) {
       await this.dropTable(table);
       throw e;
     }
   }
 
-  toColumnValue(value, _genericType) {
+  protected toColumnValue(value: string, _genericType: string): string | boolean {
```
Comment (ovr) [inline packages/cubejs-query-orchestrator/src/driver/BaseDriver.ts:298]: `GenericDataBaseType`, it's an opaque type, but we can make it as union of constants in the future.

File: packages/cubejs-query-orchestrator/src/driver/BaseDriver.ts [hunk 6]
```diff
@@ -86,17 +87,14 @@ const DbTypeValueMatcher = {
...
 };
 
-export class BaseDriver {
-  /**
-   * Workaround for Type 'BaseDriver' has no construct signatures.
-   *
-   * @param {Object} [options]
-   */
-  constructor(_options = {}) {
+export class BaseDriver implements DriverInterface {
```
Comment (ovr) [inline packages/cubejs-query-orchestrator/src/driver/BaseDriver.ts:90]: if we are going to migrate this class to TypeScript, maybe it's a good idea to make `abstract`?

File: packages/cubejs-query-orchestrator/src/driver/BaseDriver.ts [hunk 7]
```diff
@@ -172,30 +169,16 @@ export class BaseDriver {
...
 
-  /**
-   * @abstract
-   * @param {string} query
-   * @param {Array<unknown>} values
-   * @param {any} [options]
-   * @return {Promise<Array<any>>}
-   */
-  async query(_query, _values, _options = {}) {
+  public async query<R = unknown>(_query: string, _values: unknown[] = [], _options: QueryOptions = {}): Promise<R[]> {
```
Comment (ovr) [inline packages/cubejs-query-orchestrator/src/driver/BaseDriver.ts:176]: > if we are going to migrate this class to TypeScript, maybe it's a good idea to make `abstract`?

This method will be `abstract`

File: packages/cubejs-query-orchestrator/src/driver/BaseDriver.ts [hunk 8]
```diff
@@ -172,30 +169,16 @@ export class BaseDriver {
     return ssl;
   }
 
-  /**
-   * @abstract
-   */
-  async testConnection() {
+  public async testConnection() {
```
Comment (ovr) [inline packages/cubejs-query-orchestrator/src/driver/BaseDriver.ts:172]: > if we are going to migrate this class to TypeScript, maybe it's a good idea to make `abstract`?

This method will be `abstract`

File: packages/cubejs-query-orchestrator/src/driver/BaseDriver.ts [hunk 9]
```diff
@@ -86,17 +87,14 @@ const DbTypeValueMatcher = {
...
 
-export class BaseDriver {
-  /**
-   * Workaround for Type 'BaseDriver' has no construct signatures.
-   *
-   * @param {Object} [options]
-   */
-  constructor(_options = {}) {
+export class BaseDriver implements DriverInterface {
+  private logger: any;
```
Comment (ovr) [inline packages/cubejs-query-orchestrator/src/driver/BaseDriver.ts:91]: protected? Interesting to see that `DatabricksDriver` works with private field from `BaseDriver`. 

image

Comment (frolovdev) [reply packages/cubejs-query-orchestrator/src/driver/BaseDriver.ts]: Agreed

File: packages/cubejs-jdbc-driver/src/types.ts [hunk 10]
```diff
@@ -1,8 +1,13 @@
+import { Options } from 'generic-pool';
```
Comment (ovr) [inline packages/cubejs-jdbc-driver/src/types.ts:1]: import type?

**Review by ovr — APPROVED** [review]
LGTM with minor

**Pull request merged**